skyportal/models.py

```diff
@@ -2709,6 +2714,107 @@ class CommentOnSpectrum(Base, CommentMixin):
 )
 
 
+def delete_obj_if_all_data_owned(cls, user_or_token):
+    allow_nonadmins = cfg["misc.allow_nonadmins_delete_objs"] or False
+
+    deletable_photometry = Photometry.query_records_accessible_by(
+        user_or_token, mode="delete"
+    ).subquery()
+    nondeletable_photometry = (
+        DBSession()
+        .query(Photometry.obj_id)
+        .join(
+            deletable_photometry,
+            deletable_photometry.c.id == Photometry.id,
+            isouter=True,
+        )
+        .filter(deletable_photometry.c.id.is_(None))
+        .distinct(Photometry.obj_id)
+        .subquery()
+    )
+
+    deletable_spectra = Spectrum.query_records_accessible_by(
+        user_or_token, mode="delete"
+    ).subquery()
+    nondeletable_spectra = (
+        DBSession()
+        .query(Spectrum.obj_id)
+        .join(
+            deletable_spectra,
+            deletable_spectra.c.id == Spectrum.id,
+            isouter=True,
+        )
+        .filter(deletable_spectra.c.id.is_(None))
+        .distinct(Spectrum.obj_id)
+        .subquery()
+    )
+
+    deletable_candidates = Candidate.query_records_accessible_by(
+        user_or_token, mode="delete"
+    ).subquery()
+    nondeletable_candidates = (
+        DBSession()
+        .query(Candidate.obj_id)
+        .join(
+            deletable_candidates,
+            deletable_candidates.c.id == Candidate.id,
+            isouter=True,
+        )
+        .filter(deletable_candidates.c.id.is_(None))
+        .distinct(Candidate.obj_id)
+        .subquery()
+    )
+
+    deletable_sources = Source.query_records_accessible_by(
+        user_or_token, mode="delete"
+    ).subquery()
+    nondeletable_sources = (
+        DBSession()
+        .query(Source.obj_id)
+        .join(
+            deletable_sources,
+            deletable_sources.c.id == Source.id,
+            isouter=True,
+        )
+        .filter(deletable_sources.c.id.is_(None))
+        .distinct(Source.obj_id)
+        .subquery()
+    )
+
+    return (
+        DBSession()
+        .query(cls)
+        .join(
+            nondeletable_photometry,
+            nondeletable_photometry.c.obj_id == cls.id,
+            isouter=True,
+        )
+        .filter(nondeletable_photometry.c.obj_id.is_(None))
+        .join(
+            nondeletable_spectra,
+            nondeletable_spectra.c.obj_id == cls.id,
+            isouter=True,
+        )
+        .filter(nondeletable_spectra.c.obj_id.is_(None))
+        .join(
+            nondeletable_candidates,
+            nondeletable_candidates.c.obj_id == cls.id,
+            isouter=True,
+        )
+        .filter(nondeletable_candidates.c.obj_id.is_(None))
+        .join(
+            nondeletable_sources,
+            nondeletable_sources.c.obj_id == cls.id,
+            isouter=True,
+        )
+        .filter(nondeletable_sources.c.obj_id.is_(None))
+        .filter(allow_nonadmins)
```

Comment:
Does this need to be `sa.literal(allow_nonadmins)`?

Reply:
Doesn't seem like it's necessary as far as I can tell, but I added it in for clarity anyways

Reply:
Took your feedback, and also moved the permission logic code to right above the Obj class just to group things more logically.